### ISSUE TITLE
choose XSLT  processor / transformer engine?

### ISSUE THREAD

**Issue by gbanksy** [morbac/xmltools#215]
Could the transformer in use for XSLT become a user preference / option?

e.g. to be able to use Saxon-HE

Ref: https://github.com/Saxonica/Saxon-HE/ 

Reason is that the XSLT I am using is producing different results in XML Tools as compared to Saxon

The XSLT is supposed to concatenate first and last names into Full_Name

XML:

```
<?xml version='1.0' encoding='UTF-8'?><Name_Detail_Data>
  <First_Name>Percy</First_Name>
  <Last_Name>Prehn</Last_Name>
</Name_Detail_Data>
```
XSL:

```
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet version="2.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
<xsl:template match="Name_Detail_Data">
  <!-- Output the child elements (First_Name and Last_Name), including both text and tags -->
  <xsl:copy-of select="*"/>
    <!-- For the full name, output the child elements (First_Name and Last_Name) text only -->
    <Full_Name>
      <xsl:value-of select="*"/>
    </Full_Name>
  </xsl:template>
</xsl:stylesheet>
```

**Comment by martin-honnen** [comment]
@gbanksy ,

I am currently trying to extend xmltools in a fork to include SaxonC HE 12 as an XSLT 3 processor. A raw prototype works and looks as shown attached.
Would you be interested in using/testing this?
![Screenshot 2024-12-06 110900](https://github.com/user-attachments/assets/caebbaf8-d3aa-4966-9a52-5ff256c739d8)

**Comment by martin-honnen** [comment]
@morbac , would you be interested in extending your xmltools to include SaxonC HE 12 as an XSLT 3 processor (and of course also then potentially as an XQuery 3.1 and XPath 3.1 processor). Would you consider integrating changes I make in a fork into your own plugin? Or would you rather prefer to keep xmltools slim and that I rename that fork into something else/convert it into a new plugin?

**Comment by epilogic-ch** [comment]
Hi,

I haven't really had time to work on the plugin for a while and I confess I'm not using NPP as often I did before. But I'm interested in your proposal and it's in line with what I've been trying to do for some time.

In fact I'd tried to prepare the ground for this kind of option by defining the abstract class XmlWrapperInterface: the idea of this class is to be a generic interface that allows you to integrate different implementations of XML functions (MSXML, Saxon, LibXML, Xalan, etc.). The MSXMLWrapper class is for instance an implementation of this interface.

However, I still have a problem with the management of options: each XML library has specific options and currently the options specific to MSXML are integrated natively in the option window. I've still some abstraction work to be done to avoid this.

Now to answer your question: yes, it would be cool to support several xml engines and especially xslt2/3. In order to do that, the idea would be to develop a class that implements XmlWrapperInterface so it could be plugged in xmltools code easily.

**Comment by gbanksy** [comment]
Yes I’d be happy to test, although based on your screenshot it already
passes my unit test!

**Comment by martin-honnen** [comment]
@gbanksy , can you try to put the XMLTools.dll from the attached zip into the XMLTools folder of the plugins directory of your Notepad++ installation and tell me whether the Notepad++ starts up fine and doesn't complain about any module not being compatible.

Please report back your Windows version and Notepad++ version.

Make sure, of course, you keep a copy of the original XMLTools.dll in another directory to that you can revert to the original version of the plugin.

So for on two of my development machines I have had no problems using that dll but somehow when I use Windows Sandbox to work in a fresh environment where I just install Notepad++ and then add that dll Notepad++ on start gives me an error.

So I kind of would like feedback from a user whether that dll works in his environment with an existing Notepad++ installation.

Note that the only add functionality so far is XSLT 3.0 transformation from the Notepad++ menu when the active editor is an XSLT stylesheet.

[XMLTools-3.1.1.14-beta.zip](https://github.com/user-attachments/files/18059763/XMLTools-3.1.1.14-beta.zip)

**Comment by martin-honnen** [comment]
@gbanksy , I guess the zip I have posted above will not work, but I think I might have found the reason, the SaxonC dll is also needed in the XMLTools plugin folder so I have created a new zip having both dlls.
[XMLTools-3.1.1.14-beta.zip](https://github.com/user-attachments/files/18060316/XMLTools-3.1.1.14-beta.zip)

**Comment by martin-honnen** [comment]
Updated code to better handle complex XSLTs with xsl:includes and/or xsl:imports is in the attached zip.
[XMLTools-3.1.1.14-beta.zip](https://github.com/user-attachments/files/18061886/XMLTools-3.1.1.14-beta.zip)

**Comment by gbanksy** [comment]
This last zip with both dlls loaded fine into my notepad++

**Comment by alienfs** [comment]
This feature is awesome, thanks!

**Comment by martin-honnen** [comment]
I have built a new test release https://github.com/martin-honnen/xmltools/releases/tag/3.1.1.14-beta1 based on the recently published SaxonC HE 12.7 maintenance release, it works for me on a normal x64 Windows machine but as it fails for me in any attempt in a clean "Windows sandbox" environment I am not sure it will work for others.

Anyone able to test whether the above release works fine with his/her Notepad++ installation?

**Comment by alienfs** [comment]
I've tried your beta version in a Notepad++ x64 v8.8.1.

I prepared this XML file:
`<root>
  <childElement>1</childElement>
  <childElement>2</childElement>
  <childElement>3</childElement>
</root>`

I prepared this XSLT file:
`<?xml version="1.0"?>
<xsl:stylesheet version="3.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
  <xsl:output method="text"/>
  <xsl:template match="/">
    <xsl:value-of select="root/childElement"/>
  </xsl:template>
</xsl:stylesheet>`

I ran it as XSLT 3.0 Transformation... and got the expected output:
`1 2 3`

Since you weren't clear on what exactly you wanted someone else to try, that's all I did.

**Comment by martin-honnen** [comment]
Thanks, @alienfs , my main doubt was whether it works at all to run the plugin as I couldn't get it to run on a clean, "fresh" Windows like the sandbox. But I think I have resolved that by now, if the machine doesn't have a "Visual C++ Runtime redistributable" the SaxonC library and that way the plugin fails.

That is why I have updated the beta release page https://github.com/martin-honnen/xmltools/releases/tag/3.1.1.14-beta1 with [a link to the Visual C++ Runtime redistributable download](https://learn.microsoft.com/en-us/cpp/windows/latest-supported-vc-redist?view=msvc-170#latest-microsoft-visual-c-redistributable-version) and that way I hope the mere "get it to work on any Windows system, be it a developer machine or a normal user machine" is resolved.

Some helpful guys on the Notepad++ plugin forum have also pointed out that you can embed/link that runtime library directly to your plugin, I might in the future test whether that even simplifies the installation and use even more.

Thanks for your user test, happy XSLT 3.0 coding in Notepad++. Given that the main XSLT implementations like MSXML or libxslt are stuck at XSLT 1.0 I hope having an XSLT 3.0 option in simple tools like Notepad++ is a step forwards to increase the user base of XSLT 3 and to ease its use for those willing or able to leave the past in the form of XSLT 1.0 behind.

**Comment by martin-honnen** [comment]
SaxonC got an update 12.8, I have therefore updated my fork of XMLTools with XSLT 3.0 processing to that latest SaxonC HE 12.8 version: https://github.com/martin-honnen/xmltools/releases/tag/3.1.1.15-beta1.

**Comment by martin-honnen** [comment]
I have updated my fork adding XSLT 3.0 to XMLTools with the help of SaxonC HE to SaxonC HE's latest release (12.9) and have put up a release at https://github.com/martin-honnen/xmltools/releases/tag/3.1.1.16-beta1.
